### ISSUE TITLE
Device plugin code cleanup

### ISSUE THREAD

**Issue by jiayingz** [kubernetes/kubernetes#51993]
/kind feature

Here are the possible code cleanup/refactoring we have discussed during device plugin PR code review or offline:
1. Currently Kubelet side device plugin is structured as three-layers: cm.DevicePluginHandler -> deviceplugin.Manager -> deviceplugin.endpoint. We may improve code readability by dropping them to two: cm.DevicePluginHandler -> deviceplugin.Endpoint. Some of the functionalities of deviceplugin.Manager will be moved to cm.DevicePluginHandler and some may be moved to deviceplugin.Endpoint.
2. The current deviceplugin.Endpoint has two methods: list and listAndWatch that make the same gRPC call. We may merge them into one. Combined with 1) item, the deviceplugin.Endpoint interface may expose the following methods: Start(), Stop(), Alloc(), run().
- Interface cleanup: some of the interfaces in deviceplugin pkg is not necessary public and can be changed to internal.
- Most functions in deviceplugin/utils.go doesn't seem to be necessary. May consider to remove this file.

**Comment by dixudx** [comment]
@jiayingz  I would work on this. Will submit a PR later.

**Comment by lichuqiang** [comment]
@dixudx Okey...actually, I'm working on it too, have you started?

**Comment by dixudx** [comment]
@liuchuqiang Yes. Already started.

**Comment by lichuqiang** [comment]
@dixudx Okey, I'm concentrating on part of refactoring

**Comment by dixudx** [comment]
@lichuqiang I think there is no need to split this into several PRs. I'll cover it all in #52001. Thanks.

**Comment by lichuqiang** [comment]
@dixudx Well, in fact, I have finished most part of refactoring

**Comment by dixudx** [comment]
We should avoid doing duplicate work. Better notify others before starting working.

Since you've finished the main part of refactoring, I will close my pr.

**Comment by lichuqiang** [comment]
@dixudx I'm so sorry, I discussed with jiayingz in slack but forgot to response in the issue itself

**Comment by RenaudWasTaken** [comment]
Hi @jiayingz !

Sorry to be commenting so late on this.

I'll start with a short list of the Exposed interfaces / Functions of the deviceplugin module:
- Manager: `NewManagerImpl`,  `CheckpointFile`, `Start`, `Devices`, `Allocate`, `Register`, `Stop`
- Endpoint: Noting exposed
- Device Plugin Stub: `NewDevicePluginStub`, `Start`, `ListAndWatch`, `Update`, `Allocate`
- Types: `MonitorCallback`, `Manager interface`
- Utils: `IsResourceNameValid`, `IsDeviceName`


> Interface cleanup: some of the interfaces in deviceplugin pkg is not necessary public and can be changed to internal.

Not sure what interfaces you are referring to.
Looking at the interfaces exposed, `IsResourceNameValid` and `IsDeviceName` might be very similar though.
But this seem to be mostly it, am I missing something ?

> Currently Kubelet side device plugin is structured as three-layers: cm.DevicePluginHandler -> deviceplugin.Manager -> deviceplugin.endpoint. We may improve code readability by dropping them to two: cm.DevicePluginHandler -> deviceplugin.Endpoint.
> Some of the functionalities of deviceplugin.Manager will be moved to cm.DevicePluginHandler and some may be moved to deviceplugin.Endpoint.
> Combined with 1) item, the deviceplugin.Endpoint interface may expose the following methods: Start(), Stop(), Alloc(), run().

This might be a different conception or code model that we have. I do believe in separation of concerns and decoupling and the cleanest code I've read out there have been applying these.

You seem to be suggesting otherwise here by trying to merge the gRPC server (manager) in the `DevicePluginHandler` sounds a bit weird.
I already feel like the `DevicePluginHandler` is very bloated and assumes multiple roles I'm not sure if adding one more role to it is a good idea.

IMHO the current layout in the device plugin seems readable enough and follows a pretty good separation of concerns.
I'm don't feel the same of the `DevicePluginHandler` where functions seem to be assuming more than one role.

> The current deviceplugin.Endpoint has two methods: list and listAndWatch that make the same gRPC call. We may merge them into one.

I tried to wrap my head around this problem and didn't find any acceptable solutions :(
If someone can find a solution to this that would be great!

> Most functions in deviceplugin/utils.go doesn't seem to be necessary. May consider to remove this file

As mentioned, `IsResourceNameValid` might be merged with `IsDeviceName` as they seem very similar :)

However I don't see the value in increasing cognitive load on the dev reading the code by adding utils functions in the "Business Logic Code".
This seem to go against the stated objective to "improve code readability".

**Comment by vishh** [comment]
Before anyone posts a PR, let's first settle down on the right abstractions and problems to be solved. So a discussion and possibly a rough prototype will help avoid tedious code reviews.

**Comment by lichuqiang** [comment]
So I think the key point is if it necessary to combine the part of Manager into DevicePluginHandler and endpoint.
From a design perspective, it seems clearer to keep it independent; However, people might feel it redundant when looking into the code, since the manager itself does not have much logic, largely works as passthrough, except the gRPC server part.
> The current deviceplugin.Endpoint has two methods: list and listAndWatch that make the same gRPC call. We may merge them into one.

As for the merge of func listAndWatch & list. @RenaudWasTaken , I think jiayingz means that since we won't call the list func periodically as we do in the mean list/watch in k8s, it's basically a premise for func listAndWatch here,  so we may consider to combine them into one

**Comment by jiayingz** [comment]
@lichuqiang Is it possible that you break your change into small PRs? Possible one PR for a clean-up item mentioned in the issue. This way, it may be easier to settle down and also makes review faster.

**Comment by lichuqiang** [comment]
@jiayingz Sure. But as @vishh said, shall we have a discussion about direction of the  refactoring first?
First seeing the issue, I thought you major members of the feature already had a discussion offline, but it seems that RenaudWasTaken still have doubts about the idea of breaking Manager into DevicePluginHandler and endpoint.
After we reach an agreement. I would set about the modification at once :)

**Comment by jiayingz** [comment]
I think the overall goals of this issue is to improve readability and maintainability of the device plugin through certain code refactoring. The listed items are possibilities we may consider and reflects the general concerns that the reviewers had during the device plugin PR review. We do hope whoever takes this task would put more thoughts on them and come up with an implementation that helps achieve the overall goals.

In terms of reaching overall agreements on the refactoring, it sometimes could take quite a lot communication overhead. I think it is also why @vishh suggested that a rough prototype could help avoid tedious code reviews. The prototype however needs to reflect the overall goals of improving readability and maintainability (unit test coverage is part of this and shouldn't be sacrificed). If someone is seeking more specific implementation tasks, I would suggest to start with smaller items, like cleaning up functions in utils.go.

**Comment by lichuqiang** [comment]
@jiayingz Thanks for explanation, I would break the PR into small ones according to optional refactoring items :)

**Comment by lichuqiang** [comment]
As discussed, the change has been broken into small PRs as follows:

Deviceplugin refactoring: cleanup some unnecessary functions #52147
Deviceplugin refactoring: merge func list and listwatch in endpoint into one #52149

And the PR to break Manager into DevicePluginHandler and endpoint will be posted later

**Comment by RenaudWasTaken** [comment]
Hi!
Sorry for the late answers...

> From a design perspective, it seems clearer to keep it independent; However, people might feel it redundant when looking into the code, since the manager itself does not have much logic, largely works as passthrough, except for the gRPC server part.

Right, as you mentioned, it's in charge of handling `endpoints` and the
gRPC server. Furthermore it exposes a clean, readable and logical
interface.

Which seems reason enough for me to keep it as a separate module as it
would not make sense to expose this interface elsewhere.

> As for the merge of func listAndWatch & list. @RenaudWasTaken , I think jiayingz means that since we won't call the list func periodically as we do in the mean list/watch in k8s, it's basically a premise for func listAndWatch here, so we may consider to combine them into one

@lichuqiang oh this is what you meant! This was the shape of the original
function.
However IMHO shoving code in one big function does not seem like a good
programming model and seems to hamper readability.
Furthermore it made more sense to me to separate both tasks as one is listing
the devices and the other is watching them.

We can try to have list and watch call that function at the start.
I don't have a strong opinion on moving the call.
However not sure it really brings any value to merge the functions :)

> I think the overall goals of this issue is to improve readability and maintainability of the device plugin through certain code refactoring.

I think it's a good vision! I would suggest a few objectives:
- Separate the responsibilities currently handled by the device_plugin_handler before 1.8 release.
  A few tasks might be (though I'd like your opinion on this):
     - `podDevices` should it be in it's own separate file ?
     - `type DevicePluginHandler interface` shouldn't this be in the `type.go` file ?
     - should the functions be broken up in multiple ones (/files) that only accomplish one thing:
           - Resource management
           - Device Selection
           - Checkpointing (does it make sense to define an interface for it ?)
- Merge the redundant function `IsDeviceName` in `IsResourceNameValid` from utils.go before 1.8
 (I don't understand why people want to remove this file ?)
- Make `removeContents` a utils function as it does not need to be a method of the `Manager`
- Remove [this useless](https://github.com/kubernetes/kubernetes/blob/50c633182ecf188a3be156b818a9af447576ee4d/pkg/kubelet/deviceplugin/manager.go#L115) check as the socket is removed just before
- Increase the test coverage of device_plugin_handler, manager and endpoint to at least 80%
- Add e2e tests to handle the following cases:
    - Normal allocation
    - Kubelet failure
    - Device failure
    - Both failures


> The listed items are possibilities we may consider and reflects the general concerns that the reviewers had during the device plugin PR review. 

Can you point to these comments? I just went through your PR #51209 and could not find these.

**Comment by jiayingz** [comment]
On Fri, Sep 8, 2017 at 12:55 PM, Renaud Gaubert <notifications@github.com>
wrote:

> Hi!
> Sorry for the late answers...
>
> From a design perspective, it seems clearer to keep it independent;
> However, people might feel it redundant when looking into the code, since
> the manager itself does not have much logic, largely works as passthrough,
> except for the gRPC server part.
>
> Right, as you mentioned, it's in charge of handling endpoints and the
> gRPC server. Furthermore it exposes a clean, readable and logical
> interface.
>
> Which seems reason enough for me to keep it as a separate module as it
> would not make sense to expose this interface elsewhere.
>
So I can see pros and cons on either side. I think the major motivation for
dropping to two layers is to simplify the code, especially locking.


> As for the merge of func listAndWatch & list. @RenaudWasTaken
> <https://github.com/renaudwastaken> , I think jiayingz means that since
> we won't call the list func periodically as we do in the mean list/watch in
> k8s, it's basically a premise for func listAndWatch here, so we may
> consider to combine them into one
>
> @lichuqiang <https://github.com/lichuqiang> oh this is what you meant!
> This was the shape of the original
> function.
> However IMHO shoving code in one big function does not seem like a good
> programming model and seems to hamper readability.
> Furthermore it made more sense to me to separate both tasks as one is
> listing
> the devices and the other is watching them.
>
> We can try to have list and watch call that function at the start.
> I don't have a strong opinion on moving the call.
> However not sure it really brings any value to merge the functions :)
>
> I think the overall goals of this issue is to improve readability and
> maintainability of the device plugin through certain code refactoring.
>
> I think it's a good vision! I would suggest a few objectives:
>
>    - Separate the responsibilities currently handled by the
>    device_plugin_handler before 1.8 release.
>    A few tasks might be (though I'd like your opinion on this):
>       - podDevices should it be in it's own separate file ?
>
> I am fine with putting these into a separate file.

>
>    -
>       - type DevicePluginHandler interface shouldn't this be in the
>       type.go file ?
>
> The interface is only used internally in cm. I don't think we should put
it in type.go which I think should only contain public interfaces.

>
>    -
>       - should the functions be broken up in multiple ones (/files) that
>       only accomplish one thing:
>       - Resource management
>       - Device Selection
>       - Checkpointing (does it make sense to define an interface for it ?)
>
>  I think this is probably more about personal coding style. Personally, I
am not a big fan of having too many small files.
After we move podDevices functions out, the file is only about 200 line of
code, which I think is at reasonable size.

>
>    -
>       - Merge the redundant function IsDeviceName in IsResourceNameValid
>    from utils.go before 1.8
>    (I don't understand why people want to remove this file ?)
>
> 1.8 is already freeze. Only bug fixes can get in. The work we talked about
here should be for 1.9.

>
>    - Make removeContents a utils function as it does not need to be a
>    method of the Manager
>
>
>    - Remove this useless
>    <https://github.com/kubernetes/kubernetes/blob/50c633182ecf188a3be156b818a9af447576ee4d/pkg/kubelet/deviceplugin/manager.go#L115>
>    check as the socket is removed just before
>    - Increase the test coverage of device_plugin_handler, manager and
>    endpoint to at least 80%
>    - Add e2e tests to handle the following cases:
>       - Normal allocation
>       - Kubelet failure
>       - Device failure
>       - Both failures
>
> The listed items are possibilities we may consider and reflects the
> general concerns that the reviewers had during the device plugin PR review.
>
> Please take a look at PR
https://github.com/kubernetes/kubernetes/pull/51660 and issue
https://github.com/kubernetes/kubernetes/issues/52189
for e2e tests.

Jiaying

**Comment by lichuqiang** [comment]
@RenaudWasTaken Thanks for the comment.
Most of the objectives you put forward sound reasonable. In fact some of the them have been contained in existing PRs, say
 >Merge the redundant function IsDeviceName in IsResourceNameValid from utils.go before 1.8 &
 > Remove this useless check as the socket is removed just before in #52147
 > Make removeContents a utils function as it does not need to be a method of the Manager in  #52158

Since some of us still have doubts about the idea of `dropping to two layers` and `the merge of list & listwatch`, I will extract refactoring options for sure to separate PRs, and keep PR #52149 #52158 for further discussion

**Comment by RenaudWasTaken** [comment]
Hi !

> So I can see pros and cons on either side. I think the major motivation for
dropping to two layers is to simplify the code, especially locking.

In the existing PRs, I don't see a code simplification. Actually IMHO dropping an interface
is more the reverse of improving readability.

I'd rather we keep the code separated, which achieves at least separation of concern and modular
code on which we can build.

Readability is an objective but merging the layers is not achieving this and comes at the cost of removing separation of concerns and modular code.

This is not coding style but rather basic code design principles.

> The interface is only used internally in cm. I don't think we should put it in type.go which I think should only contain public interfaces.

That makes a lot of sense!

>  I think this is probably more about personal coding style. Personally, I am not a big fan of having too many small files.
> After we move podDevices functions out, the file is only about 200 line of code, which I think is at reasonable size.

File size and coding style is not the point. I'm talking of separation of concerns, not number of lines.
Having modular code that does one thing seem like a reasonable goal in terms of readability and code structure.
If we want to continue adding features everyone would be much more confortable adding them as a part of a module / struct / interface handling only the feature and not one giant structure handling everything which is what we seem to be converging to.

This argument is the same for the 2-3 layer structure. If it means that we need an additional mutex, I'm fine with that, it hardly impacts performance but on the other hand it keeps the code modular and separates concerns.
